### PR TITLE
Merging Tested Competition Changes into Main

### LINKED ISSUE #50: Robot does wheelies
Robot is doing wheelies because it is accelerating too fast. To fix this, change the closed loop ramp rate.

### PR DESCRIPTION
**Pull request by BobTheLegend**
The sum of all the changes made at the competition to create a somewhat competitive robot (Yay!)

Includes:
Fixes to belt motor problems observed before
A manual override to team color as FMS was not acting as expected
A max shooter RPM to cap the speed
Changes in camera light logic reflect the fact that nobody cared about the rules
Addition of several manual commands for use if the main "Fire" command failed
Somewhat arbitrary alterations to several constants
A new limelight ty -> RPM regression that kinda works
A cannibalized but working auto

### CODE REVIEW

File: src/main/java/frc/robot/RobotContainer.java [hunk 1]
```diff
@@ -85,13 +85,22 @@ private void configureButtonBindings() {
     final POVButton dpadUpButton = new POVButton(m_operatorXboxController, 0);
     dpadUpButton.whenPressed(new ExtendClimb(m_climb));
 
-    // Lower the launcher piston
+    // Lower the launcher pistonc
```
Comment (bjmcternan) [inline src/main/java/frc/robot/RobotContainer.java:88]: Typo

File: src/main/java/frc/robot/RobotContainer.java [hunk 2]
```diff
@@ -85,13 +85,22 @@ private void configureButtonBindings() {
...
     final POVButton dpadDownButton = new POVButton(m_operatorXboxController, 180);
     dpadDownButton.whenPressed(new RetractClimb(m_climb));
 
-    final JoystickButton xboxRotate180Btn = new JoystickButton(m_operatorXboxController,
+    final JoystickButton enableLimelight = new JoystickButton(m_operatorXboxController,
         XboxController.Button.kA.value);
-    xboxRotate180Btn.whenPressed(new Rotate(m_drivetrain, 180.0));
+    enableLimelight.whileHeld(new EnableLimelight(m_targeting));
+
```
Comment (bjmcternan) [inline src/main/java/frc/robot/RobotContainer.java:95]: Extra newline

File: src/main/java/frc/robot/RobotContainer.java [hunk 3]
```diff
@@ -110,19 +119,19 @@ public Command getAutonomousCommand() {
 
     if (allianceColor == Alliance.Blue) {
       if (DriverStation.getLocation() == 1) {
-        startingPosition = StartPositions.BLUE_LEFT;
+        startingPosition = StartPositions.BLUE_MIDDLE;
```
Comment (bjmcternan) [inline src/main/java/frc/robot/RobotContainer.java:122]: Let's not change these. If we need to force a start position we can do it on line 119. Something like this:

```
public Command getAutonomousCommand() {
    // The selected command will be run in autonomous
    // When either the alliance colour check or the location check fails it defaults
    // to the blue left side
    Command autoCommand = null;
    Alliance allianceColor = DriverStation.getAlliance();
    locationNumber = DriverStation.getLocation();


    StartPositions startingPosition = StartPositions.INVALID;

    //Force Alliance Color
   allianceColor = Alliance.Blue;

    //Force Location
    locationNumber = 2;

    if (allianceColor == Alliance.Blue) {
      if (locationNumber  == 1) {
        startingPosition = StartPositions.BLUE_MIDDLE;
      } else if (locationNumber  == 2) {
        startingPosition = StartPositions.BLUE_MIDDLE;
      } else if (locationNumber  == 3) {
        startingPosition = StartPositions.BLUE_MIDDLE;
      }
    } else if (allianceColor == Alliance.Red) {
      if (locationNumber  == 1) {
        startingPosition = StartPositions.BLUE_MIDDLE;
      } else if (locationNumber  == 2) {
        startingPosition = StartPositions.BLUE_MIDDLE;
      } else if (locationNumber  == 3) {
        startingPosition = StartPositions.BLUE_MIDDLE;
      }
    } else {
      System.out.println("WARNING - Invalid alliance color! [" + allianceColor + "]");
    }

    if (StartPositions.INVALID == startingPosition) {
      System.out.println("WARNING - Invalid starting position! [" + startingPosition + "]");
    } else {
      autoCommand = new SequentialAutoCommand(m_drivetrain, m_kinematics, m_feeder, m_targeting, m_launcher,
          startingPosition, m_climb);
    }

    return autoCommand;
  }
```

File: src/main/java/frc/robot/RobotContainer.java [hunk 4]
```diff
@@ -132,7 +141,7 @@ public Command getAutonomousCommand() {
       System.out.println("WARNING - Invalid starting position! [" + startingPosition + "]");
     } else {
       autoCommand = new SequentialAutoCommand(m_drivetrain, m_kinematics, m_feeder, m_targeting, m_launcher,
-          startingPosition);
+          startingPosition, m_climb);
```
Comment (bjmcternan) [inline src/main/java/frc/robot/RobotContainer.java:144]: Why is m_climb needed? For the compressor?

File: src/main/java/frc/robot/commands/BallControl.java [hunk 5]
```diff
@@ -14,6 +14,7 @@ public BallControl(Feeder feeder) {
 
     @Override
     public void initialize() {
+
```
Comment (bjmcternan) [inline src/main/java/frc/robot/commands/BallControl.java:17]: Extra newline

File: src/main/java/frc/robot/commands/DisableLimelight.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+    Targeting m_targeting;
+
+    public DisableLimelight(Targeting targeting) {
+        m_targeting = targeting;
+    }
+
+    @Override
+    public void initialize() {
+
```
Comment (bjmcternan) [inline src/main/java/frc/robot/commands/DisableLimelight.java:17]: Extra newline

File: src/main/java/frc/robot/commands/DisableLimelight.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
+package frc.robot.commands;
+
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.subsystems.Feeder;
+import frc.robot.subsystems.Targeting;
+
+public class DisableLimelight extends CommandBase {
+
```
Comment (bjmcternan) [inline src/main/java/frc/robot/commands/DisableLimelight.java:8]: Extra newline

File: src/main/java/frc/robot/commands/EnableLimelight.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
+package frc.robot.commands;
+
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.subsystems.Targeting;
+
+public class EnableLimelight extends CommandBase {
+
```
Comment (bjmcternan) [inline src/main/java/frc/robot/commands/EnableLimelight.java:7]: Extra newline

File: src/main/java/frc/robot/commands/EnableLimelight.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    Targeting m_targeting;
+
+    public EnableLimelight(Targeting targeting) {
+        m_targeting = targeting;
+    }
+
+    @Override
+    public void initialize() {
+
```
Comment (bjmcternan) [inline src/main/java/frc/robot/commands/EnableLimelight.java:16]: Extra newline

File: src/main/java/frc/robot/commands/Fire.java [hunk 10]
```diff
@@ -5,27 +5,31 @@
...
+        m_targeting = targeting;
         m_delayCounts = 0;
         addRequirements(m_feeder, m_launcher);
 
-        if (true == DriverStation.getAlliance().equals(Alliance.Blue)) {
-            m_teamColor = Color.kBlue;
-        } else {
-            m_teamColor = Color.kRed;
-        }
+        // if (true == DriverStation.getAlliance().equals(Alliance.Blue)) {
```
Comment (bjmcternan) [inline src/main/java/frc/robot/commands/Fire.java:27]: Why comment this out? Was the FMS not working during the competition on Saturday/Sunday?

File: src/main/java/frc/robot/commands/Fire.java [hunk 11]
```diff
@@ -101,5 +107,6 @@ public boolean isFinished() {
 
     @Override
     public void end(boolean interrupted) {
+
```
Comment (bjmcternan) [inline src/main/java/frc/robot/commands/Fire.java:110]: Extra newline

File: src/main/java/frc/robot/commands/ForceDump.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj.util.Color;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.subsystems.Launcher;
+import frc.robot.subsystems.Targeting;
+import frc.robot.Constants.LauncherConstants;
+import frc.robot.subsystems.Feeder;
+
+public class ForceDump extends CommandBase {
+
```
Comment (bjmcternan) [inline src/main/java/frc/robot/commands/ForceDump.java:14]: Extra newline

File: src/main/java/frc/robot/commands/ForceDump.java [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    @Override
+    public void execute() {
+
+        // Assume both motors will be on!
+        boolean triggerMotorOn = true;
+        boolean beltMotorOn = false;
+        boolean feed = false;
+
+        m_delayCounts--;
+
```
Comment (bjmcternan) [inline src/main/java/frc/robot/commands/ForceDump.java:48]: Extra newline

File: src/main/java/frc/robot/commands/ForceDump.java [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+        // }
+    }
+
+    @Override
+    public void initialize() {
+    }
+
+    @Override
+    public void execute() {
+
```
Comment (bjmcternan) [inline src/main/java/frc/robot/commands/ForceDump.java:41]: Extra newline

File: src/main/java/frc/robot/commands/ForceDump.java [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+    int m_delayCounts;
+
+    public ForceDump(Feeder feeder, Launcher launcher, Targeting targeting) {
+        m_feeder = feeder;
+        m_launcher = launcher;
+        m_targeting = targeting;
+        m_delayCounts = 0;
+        addRequirements(m_feeder, m_launcher);
+
+        // if (true == DriverStation.getAlliance().equals(Alliance.Blue)) {
```
Comment (bjmcternan) [inline src/main/java/frc/robot/commands/ForceDump.java:28]: If this is not needed, let's remove the commented out code

File: src/main/java/frc/robot/commands/ForceDump.java [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    @Override
+    public boolean isFinished() {
+        return false;
+    }
+
+    @Override
+    public void end(boolean interrupted) {
+
```
Comment (bjmcternan) [inline src/main/java/frc/robot/commands/ForceDump.java:96]: Extra newline

File: src/main/java/frc/robot/commands/Rotate.java [hunk 17]
```diff
@@ -2,7 +2,7 @@
 
 import edu.wpi.first.wpilibj2.command.CommandBase;
 import frc.robot.subsystems.Drivetrain;
-
+import static frc.robot.Constants.DrivetrainConstants.*;
```
Comment (bjmcternan) [inline src/main/java/frc/robot/commands/Rotate.java:5]: Let's not grab every constant. Can we just grab the drivetrain constants instead?

File: src/main/java/frc/robot/commands/SequentialAutoCommand.java [hunk 18]
```diff
@@ -56,42 +59,48 @@ public SequentialAutoCommand(Drivetrain drivetrain, Kinematics kinematics, Feede
...
                     )
                 );
                 break;
 
             case BLUE_MIDDLE:
                 addCommands(
-                    // Reset kinematics to the blue middle position 
-                    new ResetKinematics(new Position2D(-5.350, -5.363, Math.toRadians(133.5)), m_drivetrain, m_kinematics),
+                    new RetractClimb(m_climb),  
+
```
Comment (bjmcternan) [inline src/main/java/frc/robot/commands/SequentialAutoCommand.java:71]: replace this Extra newline with a comment

File: src/main/java/frc/robot/commands/SequentialAutoCommand.java [hunk 19]
```diff
@@ -56,42 +59,48 @@ public SequentialAutoCommand(Drivetrain drivetrain, Kinematics kinematics, Feede
...
                     new TrackTarget(m_drivetrain,m_targeting),
+                    //new ForceDump(m_feeder, m_launcher, m_targeting)
+                    new Fire(m_feeder, m_launcher, m_targeting)
+
                     // Fire both balls!
-                    new ParallelRaceGroup(
-                        new Fire(m_feeder, m_launcher),
-                        new Wait(3.0)
-                    ),
+                    // new ParallelRaceGroup(
```
Comment (bjmcternan) [inline src/main/java/frc/robot/commands/SequentialAutoCommand.java:85]: Let's remove this if it is not needed

File: src/main/java/frc/robot/commands/Wait.java [hunk 20]
```diff
@@ -8,12 +8,13 @@
...
         m_waitTime = waitTime;
+        m_waitTime *= 1000;
+        m_startTime = System.currentTimeMillis();
     }
 
     @Override
     public void initialize() {
-        m_waitTime *= 1000;
-        m_startTime = System.currentTimeMillis();
+
```
Comment (bjmcternan) [inline src/main/java/frc/robot/commands/Wait.java:17]: Extra newline

File: src/main/java/frc/robot/subsystems/Targeting.java [hunk 21]
```diff
@@ -64,9 +64,23 @@ public void controlLight(boolean enabled) {
 
     public double calcShooterRPM() {
         double retval = 0.0;
+        double ty = m_targetY.getDouble(0.0);
         if (m_tAcquired.getDouble(0.0) == TARGET_ACQUIRED) {
             // This is where we would calculate the distance
-            retval = DEFAULT_LAUNCHER_RPM;
+            // if(ty > 6.7){
```
Comment (bjmcternan) [inline src/main/java/frc/robot/subsystems/Targeting.java:70]: Let's remove the commented out code if it is not needed anymore

**Review by bjmcternan — CHANGES_REQUESTED** [review]
Some formatting issues and things that needs explaining

**Comment by karthikappana35** [comment]
The relevant parts of this code have been brought in DistanceToRPM which will have another PR

**Pull request closed without merging**